From 310233ded9cb6f1b77fbb5189edc20c111154fa0 Mon Sep 17 00:00:00 2001
From: Slava Egorov <vegorov@google.com>
Date: Mon, 5 Oct 2026 09:42:50 -0700
Subject: [PATCH] [vm] Retry sendfile64 on EINTR in File::Copy

Fixes https://github.com/dart-lang/sdk/issues/64473

TEST=ci

Change-Id: Ie0f8dcec8c14792b30bd78fbdf4b85f76a6a6964
Reviewed-on: https://dart-review.googlesource.com/c/sdk/+/559720
Reviewed-by: Martin Kustermann <kustermann@google.com>
Commit-Queue: Slava Egorov <vegorov@google.com>
---
 runtime/bin/file_linux.cc | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/runtime/bin/file_linux.cc b/runtime/bin/file_linux.cc
index cf8ed305..04bb047b 100644
--- a/runtime/bin/file_linux.cc
+++ b/runtime/bin/file_linux.cc
@@ -489,7 +489,9 @@ bool File::Copy(Namespace* namespc,
   intptr_t result = 1;
   while (result > 0) {
     // Loop to ensure we copy everything, and not only up to 2GB.
-    result = NO_RETRY_EXPECTED(sendfile64(new_fd, old_fd, &offset, kMaxUint32));
+    // sendfile64 can fail with EINTR if no data was written yet.
+    result =
+        TEMP_FAILURE_RETRY(sendfile64(new_fd, old_fd, &offset, kMaxUint32));
   }
   // From sendfile man pages:
   //   Applications may wish to fall back to read(2)/write(2) in the case
-- 
2.54.0

